src/com/facebook/buck/android/DexProducedFromJavaLibrary.java

```diff
@@ -228,6 +234,41 @@ public StepExecutionResult execute(ExecutionContext context) throws IOException
     return steps.build();
   }
 
+  @VisibleForTesting
+  static BuildRuleParams maybeAddDesugarDeps(
+      JavaLibrary javaLibrary, BuildRuleParams buildRuleParams) {
+    if (javaLibrary.IsDesugarEnabled()) {
+      buildRuleParams =
+          buildRuleParams.copyAppendingExtraDeps(
+              getDesugarDeps(javaLibrary).toArray(BuildRule[]::new));
+    }
+    return buildRuleParams;
+  }
+
+  @VisibleForTesting
+  static @Nullable Collection<Path> getDesugarClassPaths(
+      JavaLibrary javaLibrary, SourcePathResolver sourcePathResolver) {
+    if (!javaLibrary.IsDesugarEnabled()) {
+      // returning null will disable desugar step
+      return null;
+    }
+    return getDesugarDeps(javaLibrary)
+        .map(BuildRule::getSourcePathToOutput)
+        .filter(Objects::nonNull)
+        .map(sourcePathResolver::getAbsolutePath)
+        .collect(ImmutableSortedSet.toImmutableSortedSet(Ordering.natural()));
+  }
+
+  @VisibleForTesting
+  static Stream<BuildRule> getDesugarDeps(JavaLibrary javaLibrary) {
+    return JavaLibraryClasspathProvider.getTransitiveClasspathDeps(javaLibrary)
```

Comment:
Can we return the abi jars here? Desugaring does not require full jars. Using abi jars ensures we do not run d8 unless the abi surface has changed and should give a good build speedup